huaweicloud/services/rds/resource_huaweicloud_rds_mysql_database_privilege.go

```diff
@@ -40,233 +45,364 @@ func ResourceRdsDatabasePrivilege() *schema.Resource {
+	curArray := curJson.([]interface{})
+	rst := make([]interface{}, 0, len(curArray))
+	for _, v := range curArray {
+		rst = append(rst, map[string]interface{}{
+			"name":     utils.PathSearch("name", v, nil),
+			"readonly": utils.PathSearch("readonly", v, nil),
+		})
+	}
+	return rst
+}
 
-	return nil
+func buildGetMysqlDatabasePrivilegeQueryParams(dbName string) string {
+	return fmt.Sprintf("?db-name=%s&page=1&limit=100", dbName)
 }
 
-func resourceRdsDatabasePrivilegeDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	c := meta.(*config.Config)
-	client, err := c.HcRdsV3Client(c.GetRegion(d))
+func resourceMysqlDatabasePrivilegeUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+
+	// updateMysqlDatabasePrivilege: update RDS Mysql database privilege.
+	var (
+		updateMysqlDatabasePrivilegeProduct = "rds"
+	)
+	updateMysqlDatabasePrivilegeClient, err := cfg.NewServiceClient(updateMysqlDatabasePrivilegeProduct, region)
 	if err != nil {
 		return diag.Errorf("error creating RDS client: %s", err)
 	}
 
-	var (
-		instanceId = d.Get("instance_id").(string)
-		dbName     = d.Get("db_name").(string)
-		opts       = rds.RevokeRequestBody{
-			DbName: dbName,
-			Users:  buildRevokeUserOpts(d.Get("users").(*schema.Set).List()),
+	oldRaws, newRaws := d.GetChange("users")
+	createUsers := newRaws.(*schema.Set).List()
+	deleteUsers := buildDeleteUsers(oldRaws, newRaws)
+
+	if len(deleteUsers) > 0 {
+		err = deleteMysqlDatabasePrivilege(ctx, d, updateMysqlDatabasePrivilegeClient, deleteUsers)
+		if err != nil {
+			return diag.FromErr(err)
 		}
-		deleteReq = rds.RevokeRequest{
-			InstanceId: instanceId,
-			Body:       &opts,
+	}
+
+	if len(createUsers) > 0 {
+		err = createMysqlDatabasePrivilege(ctx, d, updateMysqlDatabasePrivilegeClient, createUsers)
+		if err != nil {
+			return diag.FromErr(err)
 		}
-	)
-	log.Printf("[DEBUG] Delete RDS database privilege options: %#v", opts)
-
-	config.MutexKV.Lock(instanceId)
-	defer config.MutexKV.Unlock(instanceId)
-	config.MutexKV.Lock(dbName)
-	defer config.MutexKV.Unlock(dbName)
-
-	err = resource.RetryContext(ctx, d.Timeout(schema.TimeoutDelete), func() *resource.RetryError {
-		_, err = client.Revoke(&deleteReq)
-		retryable, err := handleMultiOperationsError(err)
-		if retryable {
-			return resource.RetryableError(err)
+	}
+
+	return resourceMysqlDatabasePrivilegeRead(ctx, d, meta)
+}
+
+func buildDeleteUsers(oldRaws, newRaws interface{}) []interface{} {
+	oldRawUsers := oldRaws.(*schema.Set).List()
+	newRawUsers := newRaws.(*schema.Set).List()
+	newUserNames := make(map[string]bool)
+	deleteUsers := make([]interface{}, 0)
+	for _, newRawUser := range newRawUsers {
+		if raw, ok := newRawUser.(map[string]interface{}); ok {
+			newUserNames[raw["name"].(string)] = true
 		}
-		if err != nil {
-			return resource.NonRetryableError(err)
+	}
```

Comment:
```suggestion
	deletedSet  = oldRaw.(*schema.Set).Difference(newRaw.(*schema.Set))
```

Reply:
if only the param of readonly is changed, it can not be deleted, but it will be judged as deleted by this method